### PR TITLE
Revamped neighbor class

### PR DESCRIPTION
**Pull request by samwaseda**
Based on today's discussion

List of features added:
- `neigh.allow_ragged = bool` internally changes values, either filling empty slots with `inf` or removing them.
- `boundary_width_factor` and `num_neighbors_estimate_buffer` unified to `width_buffer` (sorry for bringing this chaos @niklassiemer, I hope you can live with this change...)
- `get_neighbors` and `get_neighbors_by_distance` are again virtually the same, with the only difference being the default value of `allow_ragged` (which is `True` for `get_neighbors_by_distance` and `False` for `get_neighbors`)
- `get_neighborhood` can take several points (instead of just one) and it's now based on the same tree structure as `get_neighbors`
- `neighbors.get_neighborhood` works in the same way as `structure.get_neighborhood`, with the difference being the reuse of the tree structure with `neighbors.get_neighborhood`
- Since `get_neighborhood` does not provide information on atom environment and therefore many of the methods (clustering algorithm etc.) do not make sense for `get_neighborhood`, I separated `Tree` from `Neighbors` (which is now derived from `Tree`). With `get_neighborhood`, a `Tree` instance is returned.
- `update_vectors` removed (it mostly didn't make sense anyway).

### CODE REVIEW

**Comment by samwaseda** [comment]
There are still more things to do, but since the number of lines I changed reached 100, I'd finish this PR at this point and create another one for more features.

**Review by niklassiemer — COMMENTED** [review]
Looks good to me

**Comment by samwaseda** [comment]
Since only @niklassiemer reviewed and I don't think there're too many more things to add, I decided to change this one to draft for now to do all changes I planned.

**Comment by samwaseda** [comment]
ok I have to admit I was hopelessly wrong - this one was a Pandora's box with a lot more lines to come.

**Comment by samwaseda** [comment]
So now it must be really ready. Let me know what you think.

File: pyiron/atomistics/structure/atoms.py [hunk 1]
```diff
@@ -1304,40 +1298,22 @@ def get_neighbors_by_distance(
                         (pbc are automatically taken into account)
             tolerance (int): tolerance (round decimal points) used for computing neighbor shells
             id_list (list): list of atoms the neighbors are to be looked for
-            boundary_width_factor (float): width of the layer to be added to account for pbc.
-            num_neighbors_estimate_buffer (float): Extra volume taken into account for the num_neighbors estimation
+            width_buffer (float): width of the layer to be added to account for pbc.
```
Comment (pmrv) [inline pyiron/atomistics/structure/atoms.py:1301]: Misses docs for `allow_ragged`.

File: pyiron/atomistics/structure/atoms.py [hunk 2]
```diff
@@ -1357,45 +1334,24 @@ def get_neighbors(
             tolerance (int): tolerance (round decimal points) used for computing neighbor shells
             id_list (list): list of atoms the neighbors are to be looked for
             cutoff_radius (float): Upper bound of the distance to which the search must be done
-            boundary_width_factor (float): width of the layer to be added to account for pbc.
+            width_buffer (float): width of the layer to be added to account for pbc.
```
Comment (pmrv) [inline pyiron/atomistics/structure/atoms.py:1337]: Also misses a docs for some of the arguments, should also be in the same order as the argument list.

**Review by pmrv — APPROVED** [review]
Apart from the two docstring nits, I like it!

File: pyiron/atomistics/structure/atoms.py (hunk 1 above)
Comment (samwaseda) [reply pyiron/atomistics/structure/atoms.py:1301]: How do you manage to see all the details? Do you have a software?

**Comment by liamhuber** [comment]
Sorry @samwaseda I'm not going to be able to review this in a timely manner so I cut myself from the list. I strongly approve of the idea and am confident the solution winds up looking good :)

**Comment by samwaseda** [comment]
> Sorry @samwaseda I'm not going to be able to review this in a timely manner so I cut myself from the list. I strongly approve of the idea and am confident the solution winds up looking good :)

I apologise for not properly distributing the changes to smaller PR. If I had been the reviewer, I wouldn't have wanted to read all those changes myself either.

**Comment by samwaseda** [comment]
> > Sorry @samwaseda I'm not going to be able to review this in a timely manner so I cut myself from the list. I strongly approve of the idea and am confident the solution winds up looking good :)
> 
> I apologise for not properly distributing the changes to smaller PR. If I had been the reviewer, I wouldn't have wanted to read all those changes myself either.

This being said, there was already a large amount of tests + I know @sudarsan-surendralal is a kind of person who puts new tests whenever there were preceding problems + I added new tests for all new features. So I'm rather confident that it doesn't break anything and it will do what it's supposed to do.

**Pull request merged**